test/sparse/sparse.jl

```diff
@@ -1,469 +1,518 @@
-    @test_throws DimensionMismatch permute!(A, p, q, C, Vector{eltype(A.rowval)}(length(A.colptr) - 1))
-    # Test common error checking of permute[!] methods / permutation validity
-    @test_throws ArgumentError permute!(A, (r = copy(p); r[2] = r[1]; r), q)
-    @test_throws ArgumentError permute!(A, (r = copy(p); r[2] = m + 1; r), q)
-    @test_throws ArgumentError permute!(A, p, (r = copy(q); r[2] = r[1]; r))
-    @test_throws ArgumentError permute!(A, p, (r = copy(q); r[2] = n + 1; r))
-    # Test overall functionality of [c]transpose[!] and permute[!]
-    for (m, n) in ((smalldim, smalldim), (smalldim, largedim), (largedim, smalldim))
+end
+
+@testset "conjugate" begin
+    cA = sprandn(5,5,0.2) + im*sprandn(5,5,0.2)
+    @test Array(conj.(cA)) == conj(Array(cA))
+    @test Array(conj!(copy(cA))) == conj(Array(cA))
+end
+
+@testset "SparseMatrixCSC [c]transpose[!] and permute[!]" begin
+    let smalldim = 5, largedim = 10, nzprob = 0.4
+        (m, n) = (smalldim, smalldim)
         A = sprand(m, n, nzprob)
-        At = transpose(A)
-        # transpose[!]
-        fullAt = transpose(Array(A))
-        @test transpose(A) == fullAt
-        @test transpose!(similar(At), A) == fullAt
-        # ctranspose[!]
-        C = A + im*A/2
-        fullCh = ctranspose(Array(C))
-        @test ctranspose(C) == fullCh
-        @test ctranspose!(similar(sparse(fullCh)), C) == fullCh
-        # permute[!]
+        X = similar(A)
+        C = transpose(A)
         p = randperm(m)
         q = randperm(n)
-        fullPAQ = Array(A)[p,q]
-        @test permute(A, p, q) == sparse(Array(A[p,q]))
-        @test permute!(similar(A), A, p, q) == fullPAQ
-        @test permute!(similar(A), A, p, q, similar(At)) == fullPAQ
-        @test permute!(copy(A), p, q) == fullPAQ
-        @test permute!(copy(A), p, q, similar(At)) == fullPAQ
-        @test permute!(copy(A), p, q, similar(At), similar(A.colptr)) == fullPAQ
+        @testset "common error checking of [c]transpose! methods (ftranspose!)" begin
+            @test_throws DimensionMismatch transpose!(A[:, 1:(smalldim - 1)], A)
+            @test_throws DimensionMismatch transpose!(A[1:(smalldim - 1), 1], A)
+            @test_throws ArgumentError transpose!((B = similar(A); resize!(B.rowval, nnz(A) - 1); B), A)
+            @test_throws ArgumentError transpose!((B = similar(A); resize!(B.nzval, nnz(A) - 1); B), A)
+        end
+        @testset "permute[!] error checking" begin
+            @test_throws DimensionMismatch permute(A, p[1:(end - 1)], q)
+            @test_throws DimensionMismatch permute(A, p, q[1:(end - 1)])
+            # Test common error checking of permute[!] methods / source-dest compat
+            @test_throws DimensionMismatch permute!(A[1:(m - 1), :], A, p, q)
+            @test_throws DimensionMismatch permute!(A[:, 1:(m - 1)], A, p, q)
+            @test_throws ArgumentError permute!((Y = copy(X); resize!(Y.rowval, nnz(A) - 1); Y), A, p, q)
```

Comment:
Presumably there was a reason for the "on a matrix" in the comment, e.g. to contrast with tests elsewhere for `diagm` on scalars? Preserve the additional information in the comment?